### ISSUE TITLE
GlideN64 not Matching High Res Textures Properly

### ISSUE THREAD

**Issue by ngs428** [gonetz/GLideN64#1341]
I have a few High Res texture packs installed.  They work well on Rice, but a number of the same Textures are not working in GlideN64.  GlideN64 seems to have trouble matching texture pack textures with in-game textures.

One texture in particular for example:
I am using the "Kerbs Fixed Pack  (Win32)" found at http://www.emulation64.com/files/info/951/.  

For me some of the O's show up as Q's. and the number of players select screen is seems to have a high res texture missing. I am using Mupen64plus FZ edition on my Shield. I installed the fixed pack linked above.

O and Q Issue: 
https://www.dropbox.com/s/f6pklpi8ur6n21t/2017-01-25%2022.30.18.jpg?dl=0

Player Qty Select Screen:
https://www.dropbox.com/s/ks904balscamlkt/2017-01-25%2022.29.21.jpg?dl=0

The initial splash screen of the game is also not pulling the high res textures.

Reddit thread on this: https://www.reddit.com/r/EmulationOnAndroid/comments/5qrcbu/mario_kart_64_kerbs_high_res_textures/

**Comment by Nes-player4Life** [comment]
I think this is an error within the pack itself.

**Comment by ngs428** [comment]
What type of error would it be?  It works fine in Rice.  The image missing is in the pack, it just does not display.

I don't know enough about Rice and GlideN64 to know how they handle the packs differently.

**Comment by gonetz** [comment]
@ngs428 Have you tried to enable "Alternative CRC calculation" option?

**Comment by fzurita** [comment]
Yeah, I talked to him. He tried it and it didn't seem to work. Maybe we need an alternate 2 CRC calculation.

**Comment by gonetz** [comment]
Hardly. GLideN64 reproduces original  Rice CRC method exactly. It should work unless it was changed in later versions of Rice video.

**Comment by ngs428** [comment]
As Francisco stated, I did try "alternative crc calculation". Nothing changed.  The same 2 issues I have the link on were present.  Also the Initial game screen where you have to press start, none of those textures load either.  With or without the CRC option checked.

When I switched to Rice Accurate  all those textures loaded fine.

**Comment by Nes-player4Life** [comment]
The texture itself may be misplaced. The Q was labeled as O. As for them working in rice it could be that the font textures are loaded from two locations.
I should download the pack sometime and check this theory. I do think that it is a pack issue, not plugin related.

**Comment by ngs428** [comment]
@Nes-player4Life - I installed your Mario Kart high res texture file.  Same issue on the game select screen.  The bottom section of the number of players icon is not loading properly.  

https://www.dropbox.com/s/9bo96b7ldbemehq/2017-02-10%2023.21.05.jpg?dl=0

But at least the O's and Q's work.

I tried renaming the image in Kern's fixed pack in the game select screen to the names you have in your pack.  That did not display properly either.

**Comment by ngs428** [comment]
Kerb's not Kern.

**Comment by ngs428** [comment]
Well, no matter what I tried (renaming, coping in new textures) I could not fix the issues I mentioned above.   I just went in and deleted the "1P GAME" to "4P GAME" folders and the 3 "Q" texture files.  At least there is a consistent look now without the errors.  Not all the textures show up in the game menu, but at least the in game racing looks good and there aren't any Q's where there should be O's.  

I noticed too, that the Q's that were incorrectly pulling in were the high res textures when the rest of the surrounding letters were not the high res textures.  _I too think there is a high res pack issue._

If you are interested, my fixed pack can be found here for GlideN64:
[Mario Kart 64_HIRESTEXTURES_Kerb's Fixed Pack_Nic Fix_02-11-17.zip](https://github.com/gonetz/GLideN64/files/768459/Mario.Kart.64_HIRESTEXTURES_Kerb.s.Fixed.Pack_Nic.Fix_02-11-17.zip)

**Comment by Nes-player4Life** [comment]
I'm going through the kerb pack in my free time. Hopefully We can get this resolved.

**Comment by ngs428** [comment]
I removed the Q texture files and I removed the texture files on the number of player select screen.  So both issues were corrected.  Corrected by just removing the files.

Thanks for working on this.  It is a great game that should have working textures.

**Comment by jfields687** [comment]
So I did some testing. I used [Rice's](http://www.emucr.com/2016/12/ricevideo-git-20161221.html) plugin (Works perfectly), Glide64 for PJ64 (Few minor problems), and Gonetz's GlideN64 (Loads very few textures and the Q/O incorrectly). I fixed Kerb's texture pack using Rice's plugin a couple years ago or so and remember it working very well on GlideN64 at the time so maybe there has been some changes to the CRC method since then? Gonetz, can you please compare the CRC methods to the Rice plugin if that is possible? If not, the only way to fix this is by dumping all the textures again on GlideN64 and renaming every file which would be very painful and would possibly only work for GlideN64.

**Comment by gonetz** [comment]
Sure, I'll check it, but it will hardly be soon.
Could you attach there just a little part of the problem pack, with textures, which loaded incorrect?

**Comment by jfields687** [comment]
Sure, thanks Gonetz. Here is the link to the [pack](http://www.emulation64.com/files/getfile/1313). The only textures that I saw work correctly is in the items folder. Just about all the other textures do not load or is loading incorrectly like Q/O.

**Comment by gonetz** [comment]
@jfields687 hopefully fixed: 5bfac0a66

test build:
https://drive.google.com/file/d/0B0YqMPjGo3B2N2R2Rll4NktXbWM/view?usp=sharing

please check, what is still not loaded.

**Comment by jfields687** [comment]
That did it!!! Thanks very much Gonetz!

**Comment by oddMLan** [comment]
One question. Why does it need a hack to work properly with the HD pack?

**Comment by gonetz** [comment]
> One question. Why does it need a hack to work properly with the HD pack?
So all textures become power of two?

No. Not all even numbers are power of two :)
MK64 has strange anomalies with texturing. You may see result of it as glitches in 2D elements. Another issue is that result of tile size calculation differs from the one, which expected by rice checksum algorithm. I made that hack for Glide64 and I had to apply it there.

**Comment by ngs428** [comment]
Thanks for your work on this one gonetz.  I started this issue and am using Mupen64Plus FZ on Android.  Will your changes be migrated to a released version that Francisco will merge into his app I am using?  I am not sure how the flow works from your builds to having them integrated in Mupen64Plus FZ.

Thanks again!

**Comment by fzurita** [comment]
It involves me making a new build off the app. I'll try to get one or within a week. I'm waiting on a few fixes from upstream core.

**Comment by ngs428** [comment]
Great!  Thanks Francisco!

**Comment by Gagert** [comment]
Looks like most of the textures in the Kerber texture pack works in Mario Kart 64 now, besides the minimaps in the Battle Arenas (those appears pure white while enabling this texture pack, while appearing as it should without texture packs enabled).

Also, it may be just me but it appears as there's a little bit screen tearing issues with this test build of GlideN64. The screen tearing naturally disappear while enabling the new vsync option, but I can't recall any issues with screen tearing in prior builds of GlideN64 (might be my mind playing tricks on me).

The screen tearing is easiest to spot with the spinning Nintendo logo at the startup at Mario Kart 64, and while the camera moves down the castle on the ceremony at the end of a cup.

**Comment by ngs428** [comment]
Thanks, textures are working much better now!

**Comment by gonetz** [comment]
I made several fixes in hires textures support. All issues should go. If something still not fixed, please open new issue.
